ENIGMAsystem/SHELL/Graphics_Systems/General/GSprimitives.cpp

```diff
@@ -1,17 +1,234 @@
-#include "GSprimitives.h"
-
-namespace enigma_user {
-
-unsigned draw_primitive_count(int kind, unsigned vertex_count) {
-  switch (kind) {
-    case pr_pointlist: return vertex_count;
-    case pr_linelist: return vertex_count / 2;
-    case pr_linestrip: if (vertex_count > 1) return vertex_count - 1; break;
-    case pr_trianglelist: return vertex_count / 3;
-    case pr_trianglestrip: if (vertex_count > 2) return vertex_count - 2; break;
-    case pr_trianglefan: if (vertex_count > 2) return vertex_count - 2; break;
-  }
-  return 0;
-}
-
-}
+/** Copyright (C) 2008-2013 Josh Ventura, Robert Colton
+*** Copyright (C) 2018 Robert Colton
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include "GSprimitives.h"
+#include "GSmodel.h"
+#include "GScolors.h"
+#include "GStextures.h"
+
+namespace {
+
+// the batching mode is initialized to the default here
+int draw_batch_mode = enigma_user::batch_flush_deferred;
+// the texture that was specified when the current primitive batch began
+int draw_batch_texture = -1;
+// lazy create the batch stream that we use for combining primitives
+int draw_get_batch_stream() {
+  static int draw_batch_stream = -1;
+  if (!enigma_user::d3d_model_exists(draw_batch_stream))
+    draw_batch_stream = enigma_user::d3d_model_create(enigma_user::model_stream);
+  return draw_batch_stream;
+}
+// helper function for beginning a deferred batch to determine when texture swap occurs
+void draw_batch_begin_deferred(int texId) {
+  if (draw_batch_texture != texId) {
+    enigma_user::draw_batch_flush(enigma_user::batch_flush_deferred);
+    draw_batch_texture = texId;
+  }
+}
+
+} // anonymous namespace
+
+namespace enigma_user
+{
+
+unsigned draw_primitive_count(int kind, unsigned vertex_count) {
+  switch (kind) {
+    case pr_pointlist: return vertex_count;
+    case pr_linelist: return vertex_count / 2;
+    case pr_linestrip: if (vertex_count > 1) return vertex_count - 1; break;
+    case pr_trianglelist: return vertex_count / 3;
+    case pr_trianglestrip: if (vertex_count > 2) return vertex_count - 2; break;
+    case pr_trianglefan: if (vertex_count > 2) return vertex_count - 2; break;
+  }
+  return 0;
```

Comment:
Add a debug-mode assertion fail, here (it's good to know you passed an invalid argument).

Reply:
Alright: e661c9e5f3c37f4edeb6cf17fe22dbf7e174f11b

I agree that specifying an unhandled primitive kind is an error. I did not, however, add a message for when you specify a vertex count that is too low because that's not really an error. You may have a single vertex etc and still want to know if that's enough to draw n primitives, which is the point of this function.